scipy/stats/stats.py

```diff
@@ -5670,21 +5720,143 @@ def ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate'):
+    xs = [np.swapaxes(x, axis, -1) for x in xs]
+    # determine final shape of all but the last dim
+    shape = np.broadcast(*[x[..., 0] for x in xs]).shape
+    # get final size of the dimension we're concatenating along
+    length = sum((x.shape[-1] for x in xs))
+    # create an empty array of the right shape
+    res = np.empty(shape + (length,))
+    # fill it up
+    index = 0
+    for x in xs:
+        res[..., index: index+x.shape[-1]] = x
+        index += x.shape[-1]
+    # move the axis we concatenated along to where it was
+    res = np.swapaxes(res, axis, -1)
+    return res
+
+
+def _data_permutations(data, n=-1, axis=-1, random_state=None):
+    """Vectorized permutation of data"""
+    random_state = check_random_state(random_state)
+    if axis < 0:  # we'll be adding a new dimension at the end
+        axis = data.ndim + axis
+
+    # prepare permutation indices
+    m = data.shape[axis]
+    if n > 0:
+        # replace with vectorized random_state.permutation if that gets added
+        # indices = np.argsort(random_state.random((m, n)), axis=0)
+        indices = np.array([random_state.permutation(m) for i in range(n)]).T
+    else:
+        indices = np.array(list(permutations(range(m)))).T
+    indices = indices.astype(int)
+
+    data = data.swapaxes(axis, -1)   # so we can index along a new dimension
+    data = data[..., indices]        # generate permutations
+    data = data.swapaxes(-2, axis)   # restore original axis order
+    data = np.moveaxis(data, -1, 0)  # permutations indexed along axis 0
+    return data
+
+
+def _calc_t_stat(a, b, equal_var, axis=-1):
+    """Calculate the t statistic along the given dimension"""
+    na = a.shape[axis]
+    nb = b.shape[axis]
+    avg_a = np.mean(a, axis=axis)
+    avg_b = np.mean(b, axis=axis)
+    var_a = np.var(a, axis=axis, ddof=1)
+    var_b = np.var(b, axis=axis, ddof=1)
+
+    if not equal_var:
+        denom = _unequal_var_ttest_denom(var_a, na, var_b, nb)[1]
+    else:
+        denom = _equal_var_ttest_denom(var_a, na, var_b, nb)[1]
+
+    return (avg_a-avg_b)/denom
+
+
+def _permutation_ttest(a, b, axis=0, permutations=10000, equal_var=True,
+                       random_state=None):
+    """
+    Calculates the T-test for the means of TWO INDEPENDENT samples of scores
+    using permutation methods
+
+    This test is an equivalent to `stats.ttest_ind`, except it doesn't require
+    the normality assumption since it uses a permutation test.  This function
+    is only called from ttest_ind permutations is not None.
+
+    Parameters
+    ----------
+    a, b : array_like
+        The arrays must be broadcastable, except along the dimension
+        corresponding to `axis` (the zeroth, by default).
+    axis : int, optional
+        The axis over which to operate on a and b.
+    permutations: int, optional
+        Number of permutations used to calculate p-value. If negative, use
+        all possible permutations.
```

Comment:
Not sure I saw the code for the negative case.

Reply:
In `_data_permutations`:
```
if n > 0:
    indices = np.array([random_state.permutation(m) for i in range(n)]).T
else:
    indices = np.array(list(permutations(range(m)))).T
```

Reply:
ah, `m` and `n` are different, was too late last night, thanks for clarifying.